File: app/src/main/java/com/martinlaizg/geofind/data/access/api/error/ErrorUtils.java
package com.martinlaizg.geofind.data.access.api.error;

import com.martinlaizg.geofind.data.access.api.RetrofitInstance;
import com.martinlaizg.geofind.data.access.api.service.exceptions.APIException;

import java.io.IOException;
import java.lang.annotation.Annotation;

import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Response;

public class ErrorUtils {

	/**
	 * Parse a response to a APIError
	 *
	 * @param response response to parse
	 * @return parsed response
	 */
	public static APIException parseError(Response<?> response) {
		Converter<ResponseBody, APIException> converter = RetrofitInstance.getRetrofitInstance()
				.responseBodyConverter(APIException.class, new Annotation[0]);

		APIException error = null;
		try {
			if(response.errorBody() != null) {
				error = converter.convert(response.errorBody());
			}
		} catch(IOException e) {
			return new APIException(ErrorType.PARSE);
		}

		return error;
	}
}

File: app/src/main/java/com/martinlaizg/geofind/data/access/api/error/ErrorUtils.kt
package com.martinlaizg.geofind.data.access.api.error

import com.martinlaizg.geofind.data.access.api.RetrofitFactory
import com.martinlaizg.geofind.data.access.api.service.exceptions.APIException
import retrofit2.Response
import java.io.IOException

object ErrorUtils {
	/**
	 * Parse a response to a APIException
	 *
	 * @param response response to parse
	 * @return parsed response
	 */
	fun parseError(response: Response<*>): APIException {
		val converter = RetrofitFactory.getRetrofitInstance()
				.responseBodyConverter<APIException>(APIException::class.java, arrayOfNulls(0))
		var error: APIException? = APIException(ErrorType.OTHER)
		try {
			if (response.errorBody() != null) {
				error = converter.convert(response.errorBody()) ?: error
			}
		} catch (e: IOException) {
			return APIException(ErrorType.OTHER)
		}
		return error!!
	}
}